### PR TITLE
Kernel

### PR DESCRIPTION
**Pull request by robertdown**
Ok, a few things.

This introduces a new Kernel file which is globally accessible via `$GLOBALS['kernel']`. Long term goal is for this to be the core of OpenEMR, doing all the wiring needed to modularize the system.

Currently this file is pretty thing, it only offers 2 things: a service container and an event dispatcher.

A service container allows something like this `$encounterService = $services->get('encounter')` which would give you a encounter service. Now that doesn't mean much right now because we don't actually have any services (Well, maybe 1 or 2, thanks to @MatthewVita.) The conversion is ridiculously easy, will explain at a later time.

The event dispatcher is the magical part. Let's say you've written a module which does something really cool with new patients. The thing is you have to go in and put your custom code right in that new patient form. Well that has all kinds of problems. Instead, the new event dispatcher offers a way for that new patient form to announce it's saving and for other objects to be notified and take an action. This is how we will modularize our code.

Essentially your CoolPlugin would tell the event manager "Please tell me when a new patient is saved" the event manager stores that until said action happens.

In more code driven talk...

CoolPlugin registers with the event dispatcher the event it is listening for and the method to execute. The new patient form executes the event dispather's "dispatch" method, indicating it is firing a specific event "patient.save" and passes the dispatcher the patient object. The dispatcher loops over any listener on "patient.save" and exectures the function called by CoolPlugin.

This is probably WAY easier to visualize in code, so there's a sample event subscriber and I'll create a video on it for devs. This can be a really abstract concept. Luckily, the implementation is the hard part, the end usage should be pretty straight forward!

Sorry for the novel.

### CODE REVIEW

File: library/admin/src/Event/AdminSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (robertdown) [inline library/admin/src/Event/AdminSubscriber.php]: I'll fix this

File: library/admin/src/Event/AdminSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * Created by PhpStorm.
```
Comment (robertdown) [inline library/admin/src/Event/AdminSubscriber.php]: I'll fix this

File: library/sample/src/Event/SampleEvent.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (robertdown) [inline library/sample/src/Event/SampleEvent.php]: I'll fix this to. I should really just tell PhpStorm to stop this nonsense

File: library/sample/src/Event/SampleSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (robertdown) [inline library/sample/src/Event/SampleSubscriber.php]: ... and this.

File: library/core/src/Header.php [hunk 5]
```diff
@@ -7,8 +7,10 @@
 
 namespace OpenEMR\Core;
 
+use Symfony\Component\EventDispatcher\EventDispatcher;
 use Symfony\Component\Yaml\Yaml;
 use Symfony\Component\Yaml\Exception\ParseException;
+use OpenEMR\Sample\Event\SampleEvent;
```
Comment (bradymiller) [inline library/core/src/Header.php]: just out of curiosity, why are these needed here?

**Comment by bradymiller** [comment]
i think you need to do the following composer command (noted lots of the classes being dropped in autoloader):
composer dump-autoload -o

**Comment by bradymiller** [comment]
This looks like the beginning of a nice framework. I had the benefit of @robertdown walking me through services and event managers. The plan is to get these core components into the codebase and then incorporate Esign into it as an example (it's really neat, because can modularize out the code yet still seem seamless for user; for example, even if modularize/separate the the Esign into a module, the event manager allows placing of the Esign settings in Adminstration->Globals settings like the other settings).

Before bringing into codebase, just need to ensure license/author headers and running the above composer command.

File: interface/globals.php [hunk 6]
```diff
@@ -253,6 +254,9 @@
 //    - TRACE is useful when debugging hard to spot bugs
 $GLOBALS["log_level"] = "OFF";
 
+/** @var Kernel */
+$GLOBALS["kernel"] = new Kernel();
+
 // Should Doctrine make use of connection pooling? Database connection pooling is a method
 // used to keep database connections open so they can be reused by others. (The only reason
 // to not use connection pooling is if your server has limited resources.)
```
Comment (TheToolbox) [inline interface/globals.php]: Even with limited resources, connection pooling would likely be a win.

**Comment by TheToolbox** [comment]
How does Symfony do this under the hood? It seems that either the list of handlers registered to an event would need to be stored somewhere or every single handler would need to be registered on every single page load.

File: library/core/src/Header.php (hunk 5 above)
Comment (robertdown) [inline library/core/src/Header.php]: They aren't, was a holdover from testing

**Comment by robertdown** [comment]
The Symfony event dispatcher is created for every page request, it is fairly light. It parses a YAML file which defines services that are listeners/subscribers and stores them in the event dispatcher object. The subscriptions just include the class name and method to fire so it's incredibly light and doesn't add much overhead.

There will be a core set of events that will be closely guarded, however, because of the nature of the event dispatcher, it is easy for a developer to create their own events, if the need arises.

**Comment by robertdown** [comment]
@TheToolbox Also, there's mechanisms to cache the yaml files and parsing so that we can speed up the process even more - it's in the pipeline
